orangecontrib/text/widgets/owconcordance.py

```diff
@@ -0,0 +1,265 @@
+        # indexes[0].row() == -1 indicates clicking outside of the table
+        if len(indexes) == 1 and indexes[0].row() == -1:
+            self.clear()
+            return
+        word_index = self.model().word_index
+        selected_docs = {word_index[index.row()][0] for index in indexes}
+        selected_rows = [
+            row_index for row_index, (doc_index, _) in enumerate(word_index)
+            if doc_index in selected_docs]
+        selection = QItemSelection()
+        # select all rows belonging to the selected document
+        for row in selected_rows:
+            index = self.model().index(row, 0)
+            selection.select(index, index)
+        super().select(selection, flags)
+
+
+class ConcordanceModel(QAbstractTableModel):
+    "A model for constructing concordances from text."
+
+    def __init__(self):
+        QAbstractTableModel.__init__(self)
+        self.word = None
+        self.corpus = None
+        self.indices = None
+        self.word_index = None
+        self.width = 8
+
+    def set_word(self, word):
+        self.modelAboutToBeReset.emit()
+        self.word = word
+        if self.indices is None:
+            self.word_index = None
+        else:
+            self.word_index = [
+                (doc_idx, offset) for doc_idx, doc in enumerate(self.indices)
+                 for offset in doc.offsets(self.word)]
+        self.modelReset.emit()
+
+    def set_corpus(self, corpus):
+        self.modelAboutToBeReset.emit()
+        self.corpus = corpus
+        self._compute_indices()
+        self.modelReset.emit()
+
+    def set_width(self, width):
+        self.modelAboutToBeReset.emit()
+        self.width = width
+        self.modelReset.emit()
+
+    def flags(self, _):
+        return Qt.ItemIsEnabled | Qt.ItemIsSelectable
+
+    def rowCount(self, parent=None, *args, **kwargs):
+        return 0 if parent.isValid() or self.word_index is None \
+               else len(self.word_index)
+
+    def columnCount(self, parent=None, *args, **kwargs):
+        return 3
+
+    def data(self, index, role=Qt.DisplayRole):
+        row, col = index.row(), index.column()
+        doc, index = self.word_index[row]
+
+        if role == Qt.DisplayRole:
+            tokens = self.corpus.tokens
+            if col == 0:
+                # one more if necessary to assure as much text as possible
+                # is returned
+                return ' '.join(tokens[doc][max(index - self.width, 0):index])
+            if col == 1:
+                return tokens[doc][index]
+            if col == 2:
+                return ' '.join(tokens[doc][index + 1:index + self.width + 1])
+
+        elif role == Qt.TextAlignmentRole:
+            return [Qt.AlignRight | Qt.AlignVCenter,
+                    Qt.AlignCenter,
+                    Qt.AlignLeft | Qt.AlignVCenter][col]
+
+        elif role == Qt.ForegroundRole:
+             if col == 1:
+                 return QBrush(QColor(0, 0, 255))
```

Comment:
I'm not sure about this one. I do agree the widget is already too colorful. I don't care much for the blue background, but it seems to be the best color to work with apart from grey.
I'd still prefer grey with blue foreground, but this is not possible for the previously mentioned reasons.
I'll make a wee research in the lab to see what's the opinion on GUI.